### PR TITLE
1033 filter out opted in providers

### PR DESCRIPTION
**Pull request by timabell**
### Context
Gradual migration from UCAS owned provider/course data to DfE owned data.

Update: we now have a circuit breaker in prod to reduce the risk of this change https://github.com/DFE-Digital/search-and-compare-api/pull/132

More test coverage for all the variations of opted in / out etc will be added in a subsequent PR

### Changes proposed in this pull request
* Small refactor
* TDD obeying the provider.opted_in flag in manage_courses database when importing
* Introduce builder pattern to tests in preparation for adding variation in covered scenarios.

### Guidance to review

If this breaks the importer then everything will be on :fire: 
check it carefully, run it locally, convince yourself there are no more edge cases.

To get a distribution of opted in: `update provider set opted_in=true where id % 2 = 0;`

### CODE REVIEW

**Comment by timabell** [comment]
[UCAS import should skip opted-in providers](https://trello.com/c/GyuaZt2J/1033-filter-opted-in-providers-out-of-bulk-import-process-provider-course-all-related-records)

File: src/ManageCourses.UcasCourseImporter/importer/UcasDataMigrator.cs [hunk 1]
```diff
@@ -57,20 +57,24 @@ public void UpdateUcasData()
                 }
             });
 
-            var allProviders = new Dictionary<string, Provider>();
-            MigratePerProvider("upsert providers", inst =>
+            var providerCache = _context.Providers
```
Comment (tvararu) [inline src/ManageCourses.UcasCourseImporter/importer/UcasDataMigrator.cs]: How big could this cache potentially get? Will there be enough RAM on the machine?

File: tests/ManageCourses.Tests/ImporterTests/UcasDataMigratorTests.cs [hunk 2]
```diff
@@ -116,6 +117,85 @@ public void ErroneousCourseLeavesOtherCoursesAlone()
             Context.Courses.Single().Name.Should().Be("The best title");
             Context.Courses.Single().Provider.Postcode.Should().Be(InstPostCode1);
         }
+
+        [Test]
+        public void DoesntImportOptedInProviders()
```
Comment (defong) [inline tests/ManageCourses.Tests/ImporterTests/UcasDataMigratorTests.cs]: So what happens to a course that has a accredited provider that has opted-in?

File: src/ManageCourses.UcasCourseImporter/importer/UcasDataMigrator.cs [hunk 3]
```diff
@@ -129,6 +134,11 @@ private void MigratePerProvider(string operationName, Action<UcasInstitution> ac
                 {
                     try
                     {
+                        if (providerCache.GetValueOrDefault(inst.InstCode)?.OptedIn == true)
+                        {
+                            _logger.Debug($"Skipped OptedIn provider {inst.InstCode}");
```
Comment (defong) [inline src/ManageCourses.UcasCourseImporter/importer/UcasDataMigrator.cs]: Can i have a count as well?

File: src/ManageCourses.UcasCourseImporter/importer/UcasDataMigrator.cs (hunk 1 above)
Comment (defong) [reply src/ManageCourses.UcasCourseImporter/importer/UcasDataMigrator.cs]: should be fine

File: tests/ManageCourses.Tests/ImporterTests/UcasDataMigratorTests.cs [hunk 4]
```diff
@@ -4,6 +4,7 @@
 using System.Linq;
 using Castle.Core.Logging;
 using FluentAssertions;
+using FluentAssertions.Execution;
```
Comment (defong) [inline tests/ManageCourses.Tests/ImporterTests/UcasDataMigratorTests.cs]: ErroneousCourseLeavesOtherCoursesAlone can we also have one for opted in provider as well

File: src/ManageCourses.UcasCourseImporter/importer/UcasDataMigrator.cs (hunk 3 above)
Comment (timabell) [reply src/ManageCourses.UcasCourseImporter/importer/UcasDataMigrator.cs]: good idea

**Comment by timabell** [comment]
Considering https://github.com/DFE-Digital/search-and-compare-api/pull/132 to reduce the possible impact of any bugs in this patch

File: src/ManageCourses.UcasCourseImporter/importer/UcasDataMigrator.cs (hunk 1 above)
Comment (timabell) [reply src/ManageCourses.UcasCourseImporter/importer/UcasDataMigrator.cs]: made a note on the card to check this

File: tests/ManageCourses.Tests/ImporterTests/UcasDataMigratorTests.cs (hunk 2 above)
Comment (timabell) [reply tests/ManageCourses.Tests/ImporterTests/UcasDataMigratorTests.cs]: oh

File: src/ManageCourses.UcasCourseImporter/importer/UcasDataMigrator.cs (hunk 3 above)
Comment (timabell) [reply src/ManageCourses.UcasCourseImporter/importer/UcasDataMigrator.cs]: done

File: tests/ManageCourses.Tests/ImporterTests/UcasDataMigratorTests.cs (hunk 2 above)
Comment (timabell) [reply tests/ManageCourses.Tests/ImporterTests/UcasDataMigratorTests.cs]: it's okay, we aren't deleting any providers, and you can't opt-in any that aren't in our database.

File: src/ManageCourses.UcasCourseImporter/importer/UcasDataMigrator.cs [hunk 5]
```diff
@@ -209,12 +221,10 @@ private Provider ToProvider(UcasInstitution x)
             };
         }
 
-        private Provider UpsertProvider(Provider newValues)
+        private Provider UpsertProvider(Provider newValues, IReadOnlyDictionary<string, Provider> providerCache)
```
Comment (defong) [inline src/ManageCourses.UcasCourseImporter/importer/UcasDataMigrator.cs]: Readonly collection? Don't we need to add it that collection?

File: tests/ManageCourses.Tests/ImporterTests/UcasDataMigratorTests.cs (hunk 4 above)
Comment (timabell) [reply tests/ManageCourses.Tests/ImporterTests/UcasDataMigratorTests.cs]: more test coverage is being done in a follow-up PR now

File: src/ManageCourses.UcasCourseImporter/importer/UcasDataMigrator.cs (hunk 5 above)
Comment (timabell) [reply src/ManageCourses.UcasCourseImporter/importer/UcasDataMigrator.cs]: it's done outside that method call

**Pull request merged**